### ISSUE TITLE
FP16 CUDA error: misaligned address - when using copy_attn with Apex Nvidia Amp

### ISSUE THREAD

**Issue by memray** [OpenNMT/OpenNMT-py#1475]
@francoishernandez @pltrdy 
I ran into an error after updating this commit. I ran Transformers on V100 and I tried with 1,2,4 GPUs. It works well if **I use apex_opt_level=O0** but fails with other options. It usually errors out after a few successful batches.
```
 textPayload:  "Traceback (most recent call last):
  File "/export/home/OpenNMT-summary/train.py", line 149, in run
    single_main(opt, device_id, batch_queue, semaphore)
  File "/export/home/OpenNMT-summary/onmt/train_single.py", line 145, in main
    valid_steps=opt.valid_steps)
  File "/export/home/OpenNMT-summary/onmt/trainer.py", line 243, in train
    report_stats)
  File "/export/home/OpenNMT-summary/onmt/trainer.py", line 348, in _gradient_accumulation
    report_stats.n_src_words += src_lengths.sum().item()
  File "/export/home/anaconda/envs/py37/lib/python3.7/site-packages/apex/amp/wrap.py", line 28, in wrapper
    return orig_fn(*new_args, **kwargs)
```
RuntimeError: CUDA error: misaligned address
With CUDA_LAUNCH_BLOCKING=1, it outputs:
```
Traceback (most recent call last):
  File "/export/home/OpenNMT-summary/onmt/trainer.py", line 378, in _gradient_accumulation
    self.optim.backward(loss)
  File "/export/home/OpenNMT-summary/onmt/utils/optimizers.py", line 319, in backward
    loss.backward()
  File "/export/home/anaconda/envs/py37/lib/python3.7/site-packages/torch/tensor.py", line 107, in backward
    torch.autograd.backward(self, gradient, retain_graph, create_graph)
  File "/export/home/anaconda/envs/py37/lib/python3.7/site-packages/torch/autograd/__init__.py", line 93, in backward
    allow_unreachable=True)  # allow_unreachable flag
RuntimeError: merge_sort: failed to synchronize: misaligned address
[2019-06-20 18:42:46,303 INFO] At step 5, we removed a batch - accum 3
Traceback (most recent call last):
  File "/export/home/OpenNMT-summary/train.py", line 207, in <module>
    main(opt)
  File "/export/home/OpenNMT-summary/train.py", line 93, in main
    single_main(opt, 0)
  File "/export/home/OpenNMT-summary/onmt/train_single.py", line 145, in main
    valid_steps=opt.valid_steps)
  File "/export/home/OpenNMT-summary/onmt/trainer.py", line 243, in train
    report_stats)
  File "/export/home/OpenNMT-summary/onmt/trainer.py", line 427, in _gradient_accumulation
    self.optim.step()
  File "/export/home/OpenNMT-summary/onmt/utils/optimizers.py", line 339, in step
    clip_grad_norm_(group['params'], self._max_grad_norm)
  File "/export/home/anaconda/envs/py37/lib/python3.7/site-packages/torch/nn/utils/clip_grad.py", line 32, in clip_grad_norm_
    param_norm = p.grad.data.norm(norm_type)
  File "/export/home/anaconda/envs/py37/lib/python3.7/site-packages/apex/amp/wrap.py", line 28, in wrapper
    return orig_fn(*new_args, **kwargs)
  File "/export/home/anaconda/envs/py37/lib/python3.7/site-packages/torch/tensor.py", line 253, in norm
    return torch.norm(self, p, dim, keepdim, dtype=dtype)
  File "/export/home/anaconda/envs/py37/lib/python3.7/site-packages/apex/amp/wrap.py", line 28, in wrapper
    return orig_fn(*new_args, **kwargs)
  File "/export/home/anaconda/envs/py37/lib/python3.7/site-packages/torch/functional.py", line 685, in norm
    return torch._C._VariableFunctions.norm(input, p)
RuntimeError: CUDA error: misaligned address
```

**Comment by vince62s** [comment]
Can you please post your command line for both preprocessing and training ?
thanks.

**Comment by memray** [comment]
I use my own summarization data and modified a bit the default [`config-transformer-base-1GPU.yml`](https://github.com/OpenNMT/OpenNMT-py/blob/master/config/config-transformer-base-1GPU.yml).

Now I think I found the key, it's still about the copy generator. Once I set **`copy_attn: 'true'`** it reports the above error. It happens to RNN as well. 

BTW1: Do we have a slack? for easy and quick discussion. We can start an issue only for **real issue**.
BTW2: Do we have any benchmark for the Seq2Seq models about training speed? I feel my Transformer is significantly slower than RNN.

**Comment by francoishernandez** [comment]
Quick update on this.
I am able to reproduce. It does not seem to be linked to #1465 though.
Running at commit e156cce (just before the PR), it breaks at first because of the issue you mention (need to cast batch.src_map to fp16), but when casting it I also get `CUDA error: misaligned address`.
Also tried what is suggested by @mcarilli in https://github.com/NVIDIA/apex/issues/367 --> https://github.com/francoishernandez/OpenNMT-py/commit/58d049b17b364d31129ae9d3932941c411a932cc.
For both `apex.parallel.DistributedDataParallel` as well as `torch.nn.parallel.DistributedDataParallel` I still get the same error.

**Comment by francoishernandez** [comment]
To reproduce :
```
python3 preprocess.py \
    -train_src data/src-train.txt \
    -train_tgt data/tgt-train.txt \
    -valid_src data/src-val.txt \
    -valid_tgt data/tgt-val.txt \
    -save_data data/toy_dynamic \
    -dynamic_dict
```

```
python3  train.py \
    -data data/toy_dynamic -pool_factor 1 \
    -save_model data/test_toy \
    -layers 6 -rnn_size 512 -word_vec_size 512 -transformer_ff 2048 -heads 8  \
    -encoder_type transformer -decoder_type transformer \
    -position_encoding -copy_attn  \
    -batch_size 2048 -batch_type tokens -normalization tokens -accum_count 2 \
    -max_grad_norm 0 -param_init 0  -param_init_glorot \
    -world_size 2 -gpu_ranks 0 1 \
    -optim adam -learning_rate 1e-3 \
    -model_dtype fp16
```

**Comment by vince62s** [comment]
@memray @francoishernandez 
Maybe this has been fixed with pytorch 1.2
Worth to try.

**Comment by memray** [comment]
No, same error occurs as long as copy_attn is enabled, with pytorch 1.2.

**Comment by memray** [comment]
I updated apex to nightly and pytorch to 1.3. 

With model_dtype=fp16 and apex_opt_level=O2, this problem seems to have disappeared. But I did run into `CUDA error: misaligned address` once at [L345 of trainer.py](https://github.com/OpenNMT/OpenNMT-py/blob/2a118d6273539fd2ea39c16d2687dacd60075e94/onmt/trainer.py#L345) (just stats of src_lengths, weird).

So I comment that line out and now it seems to be working fine. Will keep track of it.
